src/net_processing.cpp

```diff
@@ -4286,50 +4315,72 @@ class CompareInvMempoolOrder
-    if (MaybeDiscourageAndDisconnect(*pto)) return true;
-
-    // Don't send anything until the version handshake is complete
-    if (!pto->fSuccessfullyConnected || pto->fDisconnect)
-        return true;
-
-    // If we get here, the outgoing message serialization version is set and can't change.
-    const CNetMsgMaker msgMaker(pto->GetCommonVersion());
+    // Use mockable time for ping timeouts.
+    // This means that setmocktime may cause pings to time out.
+    auto now = GetTime<std::chrono::microseconds>();
+
+    if (gArgs.GetArg("-pingtimeout", DEFAULT_PING_TIMEOUT) &&
+        peer.m_ping_nonce_sent &&
+        now > peer.m_ping_start.load() + std::chrono::seconds{TIMEOUT_INTERVAL}) {
+        LogPrint(BCLog::NET, "ping timeout: %fs peer=%d\n", 0.000001 * count_microseconds(now - peer.m_ping_start.load()), peer.m_id);
+        node_to.fDisconnect = true;
+        return;
+    }
 
-    //
-    // Message: ping
-    //
+    const CNetMsgMaker msgMaker(node_to.GetCommonVersion());
     bool pingSend = false;
-    if (pto->fPingQueued) {
+
+    if (peer.m_ping_queued) {
         // RPC ping request by user
         pingSend = true;
     }
-    if (pto->nPingNonceSent == 0 && pto->m_ping_start.load() + PING_INTERVAL < GetTime<std::chrono::microseconds>()) {
+
+    if (peer.m_ping_nonce_sent == 0 && now > peer.m_ping_start.load() + PING_INTERVAL) {
         // Ping automatically sent as a latency probe & keepalive.
         pingSend = true;
     }
+
     if (pingSend) {
         uint64_t nonce = 0;
         while (nonce == 0) {
             GetRandBytes((unsigned char*)&nonce, sizeof(nonce));
         }
-        pto->fPingQueued = false;
-        pto->m_ping_start = GetTime<std::chrono::microseconds>();
-        if (pto->GetCommonVersion() > BIP0031_VERSION) {
-            pto->nPingNonceSent = nonce;
-            m_connman.PushMessage(pto, msgMaker.Make(NetMsgType::PING, nonce));
+        peer.m_ping_queued = false;
+        peer.m_ping_start = GetTime<std::chrono::microseconds>();
+        if (node_to.GetCommonVersion() > BIP0031_VERSION) {
+            peer.m_ping_nonce_sent = nonce;
+            m_connman.PushMessage(&node_to, msgMaker.Make(NetMsgType::PING, nonce));
         } else {
             // Peer is too old to support ping command with nonce, pong will never arrive.
-            pto->nPingNonceSent = 0;
-            m_connman.PushMessage(pto, msgMaker.Make(NetMsgType::PING));
+            peer.m_ping_nonce_sent = 0;
+            m_connman.PushMessage(&node_to, msgMaker.Make(NetMsgType::PING));
         }
     }
+}
+
+bool PeerManagerImpl::SendMessages(CNode* pto)
+{
+    // We must call MaybeDiscourageAndDisconnect first, to ensure that we'll
+    // disconnect misbehaving peers even before the version handshake is complete.
+    if (MaybeDiscourageAndDisconnect(*pto)) return true;
+
+    // Don't send anything until the version handshake is complete
```

Comment:
I know you only moved this code, but I think we could handle `fDisconnect` better. 

This comment seems misleading since I think it only refers to `fSuccesfullyConnected`. Then, we might set `fDisconnect` to true in `MaybeSendPing`, so we check this bool again a couple lines down to also return early.

A simpler logic flow might be removing this check here, starting `MaybeSendPing` with `return if node_to.fDisconnect`, and then removing this first early return from `SendMessages`. And the comment on 4377 could be updated to something along the lines of `Don't send messages to peers that are marked for disconnection`

Reply:
That seems like a reasonable change, but I'm not sure it needs to happen in this PR. Very happy to review a PR that moves this around if you think it'd make it clearer.